_posts/2019-06-04-loading-large-image-data-with-dask-array.md

```diff
@@ -67,160 +173,923 @@ pipeline.
+     width="60%">
 
 ```python
-from dask_image.imread import imread
-a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+plt.figure(figsize=(10, 10))
+skimage.io.imshow(sample[:, :, 0])
 ```
 
-If we want to handle loading some other kind of data, we can take our own
-loading function and combine it with `delayed`. This way we can handle
-operations like reading from a database or requesting data from the cloud.
+<img src="https://raw.githubusercontent.com/mrocklin/raw-host/gh-pages/images/aollsm-sample-3.png"
+     width="60%">
 
-Combining data loaded with Dask
--------------------------------
 
-Thus far we have only showed how one might load one image file. However we
-generally have many such files. At a first pass we might try to load up the
-collection of files in a list.
+### Investigate Filename Structure
+
+This are slices from only one chunk of a much larger aggregate image.
+Our main task in dealing with large image stacks is how to arrange the pieces
+back together.
+It is common to see a naming structure in the filenames. Each
+filename then may indicate a channel, time step, and spatial location with the
+`<i>` being some numeric values (possibly with units). Individual filenames may
+have more or less information and may notate it differently than we have.
+
+```
+mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif  # TODO: use filename structure from the AOLLSM example
+```
+
+In principle with NumPy we might allocate a giant array and then iteratively
+load images and place them into the giant array.
 
 ```python
-from dask_image.imread import imread
-l = [imread(fn) for fn in myfiles]
+full_array = np.empty((..., ..., ..., ..., ...), dtype=sample.dtype)
+
+for fn in filenames:
+    img = skimage.io.imread(fn)
+    index = get_location_from_filename(fn)
+    full_array[index, :, :, :] = img
 ```
 
-This is fine, but it's not that easy to manage. It would be better if we could
-turn this into one Dask Array instead of a list of many. We have a couple
-options. If we want to combine our Dask Arrays along an *existing* axis, we can
-use [`concatenate`](
-http://docs.dask.org/en/latest/array-api.html#dask.array.concatenate ).
+But in practice large image data is often too large to fit into memory, so we
+use Dask.
+
+
+### Lazily load images with Dask Array
 
+Now we learn how to lazily load and stitch together image data with Dask array.
+We'll start with simple examples first and then move onto the full example with
+this more complex dataset afterwards.
+
+We can delay the `skimage.io.imread` calls with [Dask
+Delayed](https://docs.dassk.org/en/latest/delayed.html).
 
 ```python
+import dask
 import dask.array as da
-from dask_image.imread import imread
 
-a = da.concatenate([imread(fn) for fn in myfiles])
+lazy_arrays = [dask.delayed(skimage.io.imread)(fn) for fn in filenames]
+lazy_arrays = [da.from_delayed(x, shape=sample.shape, dtype=sample.dtype)
```

Comment:
`sample` doesn't appear to be defined.

That said, from practical experience, I don't expect this to work in general. Very rarely are all of the image files uniform. At a bare minimum images near the ends may be cropped in some way. If we keep this, it would be important to note this caveat in the text above.